4.5: Data Consistency Checks

In [297]:
# Import libraries
import pandas as pd
import numpy as np
import os

Original Databases

In [298]:
#Importing data set products.csv
df_prods = pd.read_csv(r'C:\Users\Sarah\Desktop\Data Course\08-12-2022 Instacart Basket Analysis\02 Data\Original Data\products.csv', index_col = False)

Prepared Databases

In [299]:
#Importing data orders.csv from Prepared Data folder
df_ords = pd.read_csv(r'C:\Users\Sarah\Desktop\Data Course\08-12-2022 Instacart Basket Analysis\02 Data\Prepared Data\orders.csv', index_col = False)

Data Consistency Checks

In [300]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

In [301]:
del df_ords['Unnamed: 0']

In [302]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

Mixed-Type Data

In [303]:
#Create a dataframe
df_test = pd.DataFrame()

In [304]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [305]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [306]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
        print (col, ' mixed')
  else: print(col, ' consistent')

mix  mixed


In [307]:
#convert a column’s data type from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

Missing Values

In [308]:
#Finding Missing Values
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [309]:
#viewing missing values
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [310]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

Addressing Missing Values

In [311]:
df_prods.shape

(49693, 5)

In [312]:
#getting rid of missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [313]:
df_prods_clean.shape

(49677, 5)

Duplicates

In [314]:
#creating duplicate tables 
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [315]:
#finding duplicates
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

Addressing Duplicates

In [316]:
#checking rows
df_prods_clean.shape

(49677, 5)

In [317]:
df_prods_clean.drop_duplicates()

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49688       49684          Vodka, Triple Distilled, Twist of Vanilla   
49689       49685                 En Croute Roast Hazelnut Cranberry   
49690       49686                                   Artisan Baguette   
49691       49687         Smartblend Healthy Metabolism Dry Cat Food   
49692       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49688       124              5     5.3  
49689        42              1     3.1  
49690       112              3     7.8  
49691        41              8     4.7  
49692        73             11    13.5  

[49672 rows x 5 columns]

In [318]:
#excluding duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [319]:
#deleting duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [320]:
df_prods_clean_no_dups.shape

(49672, 5)

Tidying Up and Exporting Changes

In [321]:
#Exporting Prods_clean_no_dupes
#Exporting departments
df_prods_clean_no_dups.to_csv(os.path.join(r'C:\Users\Sarah\Desktop\Data Course\08-12-2022 Instacart Basket Analysis\02 Data\Prepared Data\prods_clean_no_dups.csv',))

Question 2

Run the df.describe() function on your df_prods dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.
Tip: Keep an eye on min and max values!

In [322]:
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

Because product ID, aisle ID and department ID are identifyer numbers, they don't have any actual meaning here. There seems to be an outlier in the prices with the max price being 99999, but would need to investigate further to see what the item is to see if it is a mistake or not. 

Question 3

Check for mixed-type data in your df_ords dataframe.


In [323]:
#Check for mixed types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

Question 4

If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

There were no mixed data types. 

Question 5

Run a check for missing values in your df_ords dataframe.
In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [324]:
#Finding Missing Values
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#viewing missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [325]:
#viewing missing values
df_nan_ords = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [326]:
df_ords.head(40)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   
30   2710558        3             5                   0                 17   
31   1972919        3             6                   0                 16   
32   1839752        3             7                   0                 15   
33   3225766        3             8                   0                 17   
34   3160850        3             9                   0                 16   
35    676467        3            10                   3                 16   
36    521107        3            11                   0                 18   
37   1402502        3            12                   1                 15   
38   2774568        3            13                   5                 15   
39   3343014        4             1                   6                 11   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                    

In [327]:
#checking NaN values 
df_ords.head(40) 

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   
30   2710558        3             5                   0                 17   
31   1972919        3             6                   0                 16   
32   1839752        3             7                   0                 15   
33   3225766        3             8                   0                 17   
34   3160850        3             9                   0                 16   
35    676467        3            10                   3                 16   
36    521107        3            11                   0                 18   
37   1402502        3            12                   1                 15   
38   2774568        3            13                   5                 15   
39   3343014        4             1                   6                 11   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                    

It seems that when a person makes their first order, NaN is written as they don't have any previous orders.

Question 6

Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [328]:
#Fix NaN values
df_ords['days_since_prior_order'].fillna(0, inplace=True)

In [329]:
df_ords.head(40)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   
20   1447487        2            10                   1                 11   
21   1402090        2            11                   1                 10   
22   3186735        2            12                   1                  9   
23   3268552        2            13                   4                 11   
24    839880        2            14                   3                 10   
25   1492625        2            15                   1                 11   
26   1374495        3             1                   1                 14   
27    444309        3             2                   3                 19   
28   3002854        3             3                   3                 16   
29   2037211        3             4                   2                 18   
30   2710558        3             5                   0                 17   
31   1972919        3             6                   0                 16   
32   1839752        3             7                   0                 15   
33   3225766        3             8                   0                 17   
34   3160850        3             9                   0                 16   
35    676467        3            10                   3                 16   
36    521107        3            11                   0                 18   
37   1402502        3            12                   1                 15   
38   2774568        3            13                   5                 15   
39   3343014        4             1                   6                 11   

    days_since_prior_order  
0                      0.0  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     0.0  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0  
20                     6.0  
21                    30.0  
22                    28.0  
23                    30.0  
24                    13.0  
25                    30.0  
26                    

I chose to turn the NaNs into 0 as they have not had any previous orders. 

Question 7

Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [330]:
#creating duplicate tables 
df_ords_dups = df_ords[df_ords.duplicated()]

Question 8

Address the duplicates using an appropriate method.
In a markdown cell, explain why you used your method of choice.

In [331]:
#finding duplicates
df_ords_dups.shape

(0, 6)

There are no duplicates.

Question 9

Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [332]:
#Exporting orders
df_ords.to_csv(os.path.join(r'C:\Users\Sarah\Desktop\Data Course\08-12-2022 Instacart Basket Analysis\02 Data\Prepared Data\ordersnoNaN.csv'))

In [333]:
#Exporting prods
df_prods_clean_no_dups.to_csv(os.path.join(r'C:\Users\Sarah\Desktop\Data Course\08-12-2022 Instacart Basket Analysis\02 Data\Prepared Data\prods_clean_no_dups.csv'))